trame/internal/dev/dev.py

```diff
@@ -16,40 +16,94 @@ def main():
 
     _app = get_app_instance()
     parser = _app.cli_parser
     parser.add_argument("script", help="The script to run")
-    parser.add_argument(
-        "--dev", help="Allow to dynamically reload server", action="store_true"
-    )
     args, _unknown = parser.parse_known_args()
 
-    import importlib.util
-
     spec = importlib.util.spec_from_file_location("app", args.script)
     app = importlib.util.module_from_spec(spec)
     spec.loader.exec_module(app)
 
+    setup_dev(app, clear_changes=True)
+
+    app.layout.start()
+
+
+def setup_dev(*reload_list, clear_changes=False, clear_triggers=True):
+    """
+    Set up a development environment for the trame app if --dev was passed
+    as a command line argument. If enabled, a reload button will appear at the
+    bottom of the web browser which will reload the modules that were passed as
+    arguments.
+
+    :param reload_list: positional arguments of the modules to reload when the
+                        reload button is pressed.
+    :type reload_list: python modules
+    :param clear_changes: whether or not to clear changes on reload
+    :type clear_changes: bool
+    :param clear_triggers: whether or not to clear triggers on reload
+    :type clear_triggers: bool
+    :rtype: bool
+    :returns: whether the program is running in dev mode or not
+    """
+    from trame.internal.app import get_app_instance
+
+    _app = get_app_instance()
+    parser = _app.cli_parser
+    args, _unknown = parser.parse_known_args()
+
+    if not args.dev:
+        # We are not running in development mode
+        return False
+
     # Register reload trigger
     def reload():
         print("\nReloading application...")
-        _app._change_callbacks.clear()
-        _app._triggers.clear()
+
+        if clear_changes:
+            _app._change_callbacks.clear()
+
+        if clear_triggers:
+            # Never clear these triggers
+            trigger_whitelist = [
+                "js_error",
+                "server_reload",
+            ]
+            _remove_keys(_app._triggers, trigger_whitelist)
+
         _app.reload_app()
 
-        # Keep sys trame ones
-        _app._triggers["server_reload"] = reload
-        _app._triggers["js_error"] = log_js_error
+        def reload_modules(*modules):
+            for m in modules:
+                m.__loader__.exec_module(m)
+
+                if hasattr(m, 'on_reload'):
+                    m.on_reload(reload_modules)
+
+        for module in reload_list:
+            reload_modules(module)
+
+            if hasattr(module, 'layout'):
+                module.layout.flush_content()
 
-        spec = importlib.util.spec_from_file_location("app", args.script)
-        app = importlib.util.module_from_spec(spec)
-        spec.loader.exec_module(app)
-        app.layout.flush_content()
         print(" > done !\n")
 
-    _app.trigger("server_reload")(reload)
+    # Set the reload trigger
+    _app._triggers["server_reload"] = reload
```

Comment:
Do you prefer hat syntax vs the decorator call?

Reply:
Decorator is probably better. I'll fix it.